openedx/core/djangoapps/content/block_structure/models.py

```diff
@@ -59,6 +61,63 @@ def _bs_model_storage():
     )
 
 
+class CustomizableFileField(models.FileField):
+    """
+    Subclass of FileField that allows custom settings to not
+    be serialized (hard-coded) in migrations. Otherwise,
+    migrations include optional settings for storage (such as
+    the storage class and bucket name); we don't want to
+    create new migration files for each configuration change.
+    """
+    def __init__(self, *args, **kwargs):
+        kwargs.update(dict(
+            upload_to=_path_name,
+            storage=_bs_model_storage(),
+            max_length=500,  # allocate enough for base path + prefix + usage_key + timestamp in filepath
+        ))
+        super(CustomizableFileField, self).__init__(*args, **kwargs)
+
+    def deconstruct(self):
+        name, path, args, kwargs = super(CustomizableFileField, self).deconstruct()
+        del kwargs['upload_to']
+        del kwargs['storage']
+        del kwargs['max_length']
+        return name, path, args, kwargs
+
+
+@contextmanager
+def _storage_error_handling(bs_model, operation, is_read_operation=False):
+    """
+    Helpful context manager that handles various errors
+    from the backend storage.
+
+    Typical errors at read time on configuration changes:
+        IOError:
+            - File not found (S3 or FS)
+            - Bucket name changed (S3)
+        SuspiciousOperation
+            - Path mismatches when changing backends
+
+    Other known errors:
+        OSError
+            - Access issues in creating files (FS)
+        S3ResponseError
+            - Incorrect credentials with 403 status (S3)
+            - Non-existent bucket with 404 status (S3)
+    """
+    try:
+        yield
+    except Exception as error:  # pylint: disable=broad-except
+        log.exception(u'BlockStructure: Exception %s on store %s; %s.', error.__class__, operation, bs_model)
```

Comment:
**Note to reviewers:** I added log statements for all storage-based issues so it's now easier to find and debug in Splunk, by searching for the "BlockStructure" prefix.